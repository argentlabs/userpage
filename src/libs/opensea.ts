import joinUrl from "url-join"

export interface OpenSeaResponse {
  assets: AssetElement[]
}

export interface AssetElement {
  id: number
  token_id: string
  num_sales: number
  background_color: null
  image_url: string
  image_preview_url: string
  image_thumbnail_url: string
  image_original_url: null | string
  animation_url: null
  animation_original_url: null
  name: string
  description: null | string
  external_link: null | string
  asset_contract: AssetContract
  permalink: string
  collection: Collection
  decimals: number | null
  token_metadata: null | string
  owner: Creator
  sell_orders: SellOrder[] | null
  creator: Creator
  traits: Trait[]
  last_sale: null
  top_bid: null
  listing_date: null
  is_presale: boolean
  transfer_fee_payment_token: null
  transfer_fee: null
}

export interface AssetContract {
  address: string
  asset_contract_type: string
  created_date: Date
  name: string
  nft_version: null | string
  opensea_version: null | string
  owner: number
  schema_name: string
  symbol: string
  total_supply: null | string
  description: null | string
  external_link: null | string
  image_url: null | string
  default_to_fiat: boolean
  dev_buyer_fee_basis_points: number
  dev_seller_fee_basis_points: number
  only_proxied_transfers: boolean
  opensea_buyer_fee_basis_points: number
  opensea_seller_fee_basis_points: number
  buyer_fee_basis_points: number
  seller_fee_basis_points: number
  payout_address: null
}

export interface Collection {
  banner_image_url: null | string
  chat_url: null
  created_date: Date
  default_to_fiat: boolean
  description: null | string
  dev_buyer_fee_basis_points: string
  dev_seller_fee_basis_points: string
  discord_url: null | string
  display_data: DisplayData
  external_url: null | string
  featured: boolean
  featured_image_url: null | string
  hidden: boolean
  safelist_request_status: string
  image_url: null | string
  is_subject_to_whitelist: boolean
  large_image_url: null | string
  medium_username: null | string
  name: string
  only_proxied_transfers: boolean
  opensea_buyer_fee_basis_points: string
  opensea_seller_fee_basis_points: string
  payout_address: null | string
  require_email: boolean
  short_description: null
  slug: string
  telegram_url: null | string
  twitter_username: null | string
  instagram_username: null | string
  wiki_url: null
}

export interface DisplayData {
  card_display_style: string
  images?: any[]
}

export interface Creator {
  user: User | null
  profile_img_url: string
  address: string
  config: string
}

export interface User {
  username: null | string
}

export interface SellOrder {
  created_date: Date
  closing_date: null
  closing_extendable: boolean
  expiration_time: number
  listing_time: number
  order_hash: string
  metadata: Metadata
  exchange: string
  maker: FeeRecipient
  taker: FeeRecipient
  current_price: string
  current_bounty: string
  bounty_multiple: string
  maker_relayer_fee: string
  taker_relayer_fee: string
  maker_protocol_fee: string
  taker_protocol_fee: string
  maker_referrer_fee: string
  fee_recipient: FeeRecipient
  fee_method: number
  side: number
  sale_kind: number
  target: string
  how_to_call: number
  calldata: string
  replacement_pattern: string
  static_target: string
  static_extradata: string
  payment_token: string
  payment_token_contract: PaymentTokenContract
  base_price: string
  extra: string
  quantity: string
  salt: string
  v: number
  r: string
  s: string
  approved_on_chain: boolean
  cancelled: boolean
  finalized: boolean
  marked_invalid: boolean
  prefixed_hash: string
}

export interface FeeRecipient {
  user: number
  profile_img_url: string
  address: string
  config: string
}

export interface Metadata {
  asset: MetadataAsset
  schema: string
}

export interface MetadataAsset {
  id: string
  address: string
  quantity: string
}

export interface PaymentTokenContract {
  id: number
  symbol: string
  address: string
  image_url: string
  name: string
  decimals: number
  eth_price: string
  usd_price: string
}

export interface Trait {
  trait_type: string
  value: string
  display_type: null
  max_value: null
  trait_count: number
  order: null
}

export const getNftMediaUrl = (
  nft?: AssetElement,
  detailView: boolean = false,
): string =>
  detailView
    ? nft?.animation_original_url ||
      nft?.animation_url ||
      nft?.image_original_url ||
      nft?.image_url ||
      nft?.image_preview_url ||
      nft?.image_thumbnail_url ||
      "error"
    : nft?.animation_url ||
      nft?.animation_original_url ||
      nft?.image_url ||
      nft?.image_original_url ||
      nft?.image_preview_url ||
      nft?.image_thumbnail_url ||
      "error"

const {
  REACT_APP_OPENSEA_ENDPOINT = "https://api.opensea.io/api/v1/assets",
  REACT_APP_OPENSEA_API_KEY,
} = process.env

export const fetchNfts = async (address: string): Promise<AssetElement[]> => {
  const response = await fetch(
    joinUrl(
      REACT_APP_OPENSEA_ENDPOINT,
      `?owner=${address}&order_direction=desc&offset=0&limit=50`,
    ),
    {
      headers: {
        ...(REACT_APP_OPENSEA_API_KEY && {
          "X-API-KEY": REACT_APP_OPENSEA_API_KEY,
        }),
      },
    },
  )

  if (response.status === 400) throw Error("Wrong address")
  if (response.status >= 400) throw Error("Request failed")

  const nfts = (await response.json()) as OpenSeaResponse

  return nfts.assets
}

export const fetchNft = async (
  tokenId: string,
  assetContractAddress: string,
): Promise<AssetElement> => {
  const response = await fetch(
    joinUrl(
      REACT_APP_OPENSEA_ENDPOINT,
      `?token_ids=${tokenId}&asset_contract_address=${assetContractAddress}&order_direction=desc&offset=0&limit=1`,
    ),
    {
      headers: {
        ...(REACT_APP_OPENSEA_API_KEY && {
          "X-API-KEY": REACT_APP_OPENSEA_API_KEY,
        }),
      },
    },
  )

  if (response.status === 400) throw Error("Wrong address")
  if (response.status >= 400) throw Error("Request failed")

  const nfts = (await response.json()) as OpenSeaResponse

  return nfts.assets[0]
}
