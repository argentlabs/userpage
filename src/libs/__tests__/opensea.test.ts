import fetchMock from "jest-fetch-mock"

import { AssetElement, OpenSeaResponse, fetchNfts } from "../opensea"

describe("[unit] Opensea", () => {
  beforeAll(() => {
    fetchMock.enableMocks()
  })
  beforeEach(() => {
    fetchMock.resetMocks()
  })
  afterAll(() => {
    fetchMock.disableMocks()
  })
  it("transforms 200", async () => {
    const mockResponse: OpenSeaResponse = {
      assets: [],
    }
    const expectedResult: AssetElement[] = []
    fetchMock.mockResponseOnce(JSON.stringify(mockResponse))

    const response = await fetchNfts("address")

    expect(response).toStrictEqual(expectedResult)
  })
  it("throws on 404", async () => {
    fetchMock.mockOnce(async () => ({ status: 400, body: "" }))

    await expect(fetchNfts("janek")).rejects.toThrowError("Wrong address")
  })
  it("throws on 500", async () => {
    fetchMock.mockOnce(async () => ({ status: 500, body: "" }))

    await expect(fetchNfts("janek")).rejects.toThrowError("Request failed")
  })
})

describe("[int] Opensea", () => {
  it("works with '0xe75AFa985de4F013AccFc8b4Dd744551C6EEB5a9'", async () => {
    const fetchReal = fetch

    fetchMock.enableMocks()

    let lastResponseClone: Response | undefined
    const fetchJestFn = jest.fn<
      Promise<Response>,
      [input: RequestInfo, init?: RequestInit | undefined]
    >((...args) =>
      fetchReal(...args).then((x) => {
        lastResponseClone = x.clone()
        return x
      }),
    )
    fetchMock.mockImplementation(fetchJestFn as any)

    const response = await fetchNfts(
      "0xe75AFa985de4F013AccFc8b4Dd744551C6EEB5a9",
    )

    expect(response.map((x) => x.id)).toMatchSnapshot("fetchNfts")
    // fetch gets called once
    expect(fetchJestFn.mock.calls.length).toBe(1)
    // calls the right url
    expect(fetchJestFn.mock.calls[0][0]).toEqual(
      "https://api.opensea.io/api/v1/assets?owner=0xe75AFa985de4F013AccFc8b4Dd744551C6EEB5a9&order_direction=desc&offset=0&limit=50",
    )
    // gets correct api response
    const fetchResponse: Response = await fetchJestFn.mock.results[0].value
    expect(fetchResponse.status).toBe(200)
    await expect(lastResponseClone?.json()).resolves.toMatchSnapshot(
      "fetchNftsResponse",
    )

    fetchMock.disableMocks()
  })
  it("doesnt work with wrong ans", async () => {
    const fetchReal = fetch

    fetchMock.enableMocks()

    const fetchJestFn = jest.fn(fetchReal)
    fetchMock.mockImplementation(fetchJestFn as any)

    await expect(fetchNfts("0x0")).rejects.toThrowError("Wrong address")

    // fetch gets called once
    expect(fetchJestFn.mock.calls.length).toBe(1)
    // calls the right url
    expect(fetchJestFn.mock.calls[0][0]).toEqual(
      "https://api.opensea.io/api/v1/assets?owner=0x0&order_direction=desc&offset=0&limit=50",
    )
    // gets correct api response
    const fetchResponse: Response = await fetchJestFn.mock.results[0].value
    expect(fetchResponse.status).toBe(400)

    fetchMock.disableMocks()
  })
})
