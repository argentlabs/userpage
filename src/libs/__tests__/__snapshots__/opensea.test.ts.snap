// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[int] Opensea works with '0xe75AFa985de4F013AccFc8b4Dd744551C6EEB5a9': fetchNfts 1`] = `
Array [
  25404240,
]
`;

exports[`[int] Opensea works with '0xe75AFa985de4F013AccFc8b4Dd744551C6EEB5a9': fetchNftsResponse 1`] = `
Object {
  "assets": Array [
    Object {
      "animation_original_url": null,
      "animation_url": null,
      "asset_contract": Object {
        "address": "0x57f1887a8bf19b14fc0df6fd9b2acc9af147ea85",
        "asset_contract_type": "non-fungible",
        "buyer_fee_basis_points": 0,
        "created_date": "2019-05-08T21:59:29.327544",
        "default_to_fiat": false,
        "description": "Ethereum Name Service (ENS) domains are secure domain names for the decentralized world. ENS domains provide a way for users to map human readable names to blockchain and non-blockchain resources, like Ethereum addresses, IPFS hashes, or website URLs. ENS domains can be bought and sold on secondary markets.",
        "dev_buyer_fee_basis_points": 0,
        "dev_seller_fee_basis_points": 0,
        "external_link": "https://ens.domains",
        "image_url": "https://lh3.googleusercontent.com/0cOqWoYA7xL9CkUjGlxsjreSYBdrUBE0c6EO1COG4XE8UeP-Z30ckqUNiL872zHQHQU5MUNMNhfDpyXIP17hRSC5HQ=s60",
        "name": "ENS",
        "nft_version": "3.0",
        "only_proxied_transfers": false,
        "opensea_buyer_fee_basis_points": 0,
        "opensea_seller_fee_basis_points": 250,
        "opensea_version": null,
        "owner": 279872,
        "payout_address": null,
        "schema_name": "ERC721",
        "seller_fee_basis_points": 250,
        "symbol": "ENS",
        "total_supply": null,
      },
      "background_color": null,
      "collection": Object {
        "banner_image_url": null,
        "chat_url": null,
        "created_date": "2019-05-08T21:59:36.282454",
        "default_to_fiat": false,
        "description": "Ethereum Name Service (ENS) domains are secure domain names for the decentralized world. ENS domains provide a way for users to map human readable names to blockchain and non-blockchain resources, like Ethereum addresses, IPFS hashes, or website URLs. ENS domains can be bought and sold on secondary markets.",
        "dev_buyer_fee_basis_points": "0",
        "dev_seller_fee_basis_points": "0",
        "discord_url": null,
        "display_data": Object {
          "card_display_style": "cover",
        },
        "external_url": "https://ens.domains",
        "featured": false,
        "featured_image_url": "https://lh3.googleusercontent.com/BBj09xD7R4bBtg1lgnAAS9_TfoYXKwMtudlk-0fVljlURaK7BWcARCpkM-1LGNGTAcsGO6V1TgrtmQFvCo8uVYW_QEfASK-9j6Nr=s300",
        "hidden": false,
        "image_url": "https://lh3.googleusercontent.com/0cOqWoYA7xL9CkUjGlxsjreSYBdrUBE0c6EO1COG4XE8UeP-Z30ckqUNiL872zHQHQU5MUNMNhfDpyXIP17hRSC5HQ=s60",
        "instagram_username": null,
        "is_subject_to_whitelist": false,
        "large_image_url": "https://lh3.googleusercontent.com/BBj09xD7R4bBtg1lgnAAS9_TfoYXKwMtudlk-0fVljlURaK7BWcARCpkM-1LGNGTAcsGO6V1TgrtmQFvCo8uVYW_QEfASK-9j6Nr=s300",
        "medium_username": "the-ethereum-name-service",
        "name": "ENS: Ethereum Name Service",
        "only_proxied_transfers": false,
        "opensea_buyer_fee_basis_points": "0",
        "opensea_seller_fee_basis_points": "250",
        "payout_address": null,
        "require_email": false,
        "safelist_request_status": "verified",
        "short_description": null,
        "slug": "ens",
        "telegram_url": null,
        "twitter_username": "ensdomains",
        "wiki_url": null,
      },
      "creator": Object {
        "address": "0x4fe4e666be5752f1fdd210f4ab5de2cc26e3e0e8",
        "config": "",
        "profile_img_url": "https://storage.googleapis.com/opensea-static/opensea-profile/9.png",
        "user": null,
      },
      "decimals": null,
      "description": "\`rahrt.eth\`, an ENS name.

Please check the expiration date. To keep your name beyond that date, you will need to pay a renewal fee, currently set at $5/year for names 5 characters or longer, $160/year for names 4 characters in length, and $640/year for names 3 characters in length.",
      "external_link": "https://app.ens.domains/search/rahrt",
      "id": 25404240,
      "image_original_url": null,
      "image_preview_url": "https://lh3.googleusercontent.com/W7PWHRBknZdrjgNt9dSNCWdlraA41Dw9PhbiVqO2BNTaXfW5AXhOAOLYctDNgVHbma2hvuZsYiy4UFCVsMf-J8yokCCTkyPC3U-wdEc=s250",
      "image_thumbnail_url": "https://lh3.googleusercontent.com/W7PWHRBknZdrjgNt9dSNCWdlraA41Dw9PhbiVqO2BNTaXfW5AXhOAOLYctDNgVHbma2hvuZsYiy4UFCVsMf-J8yokCCTkyPC3U-wdEc=s128",
      "image_url": "https://lh3.googleusercontent.com/W7PWHRBknZdrjgNt9dSNCWdlraA41Dw9PhbiVqO2BNTaXfW5AXhOAOLYctDNgVHbma2hvuZsYiy4UFCVsMf-J8yokCCTkyPC3U-wdEc",
      "is_presale": false,
      "last_sale": null,
      "listing_date": null,
      "name": "rahrt.eth",
      "num_sales": 0,
      "owner": Object {
        "address": "0xe75afa985de4f013accfc8b4dd744551c6eeb5a9",
        "config": "",
        "profile_img_url": "https://storage.googleapis.com/opensea-static/opensea-profile/21.png",
        "user": null,
      },
      "permalink": "https://opensea.io/assets/0x57f1887a8bf19b14fc0df6fd9b2acc9af147ea85/13334487929672912875225218455723364774141570484356505357480240172131791493081",
      "sell_orders": null,
      "token_id": "13334487929672912875225218455723364774141570484356505357480240172131791493081",
      "token_metadata": null,
      "top_bid": null,
      "traits": Array [
        Object {
          "display_type": null,
          "max_value": null,
          "order": null,
          "trait_count": 62684,
          "trait_type": "tag",
          "value": "Short Name",
        },
        Object {
          "display_type": "number",
          "max_value": null,
          "order": null,
          "trait_count": 0,
          "trait_type": "Length",
          "value": 5,
        },
      ],
      "transfer_fee": null,
      "transfer_fee_payment_token": null,
    },
  ],
}
`;
